automates/program_analysis/CAST2GrFN/visitors/cast_to_air_visitor.py

```diff
@@ -37,197 +46,1452 @@
+            # has no outputs
+            cond_con.add_condition_outputs(-1, [])
+
+        # Just the name of all currently defined vars
+        external_vars = [
+            v.identifier_information.name
+            for v in self.state.variables
+            if v.identifier_information.scope[: len(self.state.scope_stack)]
+            != self.state.scope_stack
+        ]
+
+        # Gather the maximum updated version of variables for variables that
+        # were defined outside of this if block
+        max_output_vars = {}
+        for b in body_result:
+            for v in b.output_variables:
+                name = v.identifier_information.name
+                if name in external_vars and (
+                    name not in max_output_vars
+                    or v.version > max_output_vars[name].version
+                ):
+                    max_output_vars[name] = v
+        callee_output_vars = list(max_output_vars.values())
+
+        if condition_type == "LOOP":
+            # Create exit condition variable assign
+            exit_expr_var_name = f"EXIT"
+            decision_assign = Assignment(
+                left=Var(val=Name(name=exit_expr_var_name), type="Boolean"),
+                right=UnaryOp(
+                    op=UnaryOperator.NOT,
+                    value=Name(name=cond_expr_var_name),
+                    source_refs=expr.source_refs,
+                ),
+                source_refs=expr.source_refs,
+            )
+            decision_assign_result = self.visit(decision_assign)[-1]
+            decision_assign_result.output_variables[-1].add_metadata(
+                generate_from_source_metadata(False, "LOOP_EXIT_VAR")
+            )
+            # Update type of lambda to decision
+            decision_assign_result = C2AExpressionLambda(
+                decision_assign_result.identifier_information,
+                decision_assign_result.input_variables,
+                decision_assign_result.output_variables,
+                [],
+                C2ALambdaType.DECISION,
+                decision_assign_result.source_ref,
+                decision_assign_result.lambda_expr,
+                # TODO actually fill out ast
+                decision_assign_result.cast,
+            )
+
+            body_result.append(decision_assign_result)
+
+            # We have to pass each input var through the
+            matching_vars = [
+                (iv, ov)
+                for iv in cond_con.arguments
+                for ov in callee_output_vars
+                if ov.identifier_information.name == iv.identifier_information.name
+            ]
+            if matching_vars:
+                decision_var_names = [
+                    v[1].identifier_information.name for v in matching_vars
+                ]
+
+                vars_to_output_from_input_decision = list()
+                for b in body_result:
+
+                    def enumerate_vars_and_update_version(vars):
```

Comment:
@dpdicken I don't see any usage of the loop iterator variable `b` referenced directly by name in this function. Can you please move its definition out from under the loop scope?